Kernel/FileSystem/InodeWatcher.cpp

```diff
@@ -62,37 +70,87 @@ KResultOr<size_t> InodeWatcher::read(FileDescription&, u64, UserOrKernelBuffer&
     return bytes_to_write;
 }
 
-KResultOr<size_t> InodeWatcher::write(FileDescription&, u64, const UserOrKernelBuffer&, size_t)
+KResult InodeWatcher::close()
 {
-    return EIO;
+    Locker locker(m_lock);
+
+    for (auto& entry : m_wd_to_watches) {
+        // NOTE: see comment in unregister_inode.
+        auto& inode = const_cast<Inode&>(entry.value->inode);
+        inode.unregister_watcher({}, *this);
+    }
+
+    m_wd_to_watches.clear();
+    m_inode_to_watches.clear();
+    return KSuccess;
 }
 
 String InodeWatcher::absolute_path(const FileDescription&) const
 {
-    if (auto inode = m_inode.strong_ref())
-        return String::formatted("InodeWatcher:{}", inode->identifier().to_string());
-    return "InodeWatcher:(gone)";
+    return String::formatted("InodeWatcher:({})", m_wd_to_watches.size());
 }
 
-void InodeWatcher::notify_inode_event(Badge<Inode>, InodeWatcherEvent::Type event_type)
+void InodeWatcher::notify_inode_event(Badge<Inode>, InodeIdentifier inode_id, InodeWatcherEvent::Type event_type, String const& name)
 {
     Locker locker(m_lock);
-    m_queue.enqueue({ event_type });
+
+    auto it = m_inode_to_watches.find(inode_id);
+    if (it == m_inode_to_watches.end())
+        return;
+
+    auto& watcher = *it->value;
+    if (!(watcher.event_mask & static_cast<unsigned>(event_type)))
+        return;
+
+    m_queue.enqueue({ watcher.wd, event_type, name });
     evaluate_block_conditions();
 }
 
-void InodeWatcher::notify_child_added(Badge<Inode>, const InodeIdentifier& child_id)
+KResultOr<int> InodeWatcher::register_inode(Inode& inode, unsigned event_mask)
 {
     Locker locker(m_lock);
-    m_queue.enqueue({ InodeWatcherEvent::Type::ChildAdded, child_id.index().value() });
-    evaluate_block_conditions();
+
+    if (m_inode_to_watches.find(inode.identifier()) != m_inode_to_watches.end())
+        return EEXIST;
+
+    int wd = m_wd_counter++;
```

Comment:
What happens if `m_wd_counter` wraps around?

Reply:
Pretty sure we'd hit `EMFILE` before reaching 2.1 billion open watches :^) What should be here instead? I think I can use a `Checked` here and enforce wraparound for that unlikely case.

Reply:
If you open and close watchers repeatedly, you will eventually hit this without `EMFILE` getting in your way. Or am I missing something>

Reply:
Yes you will, I've made it use a `Checked<int>` now.